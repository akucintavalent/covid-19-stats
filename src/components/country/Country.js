import { useParams } from 'react-router-dom';

const Country = () => {
  const { country } = useParams();
  return (
    <div>{country}</div>
  );
};

export default Country;
