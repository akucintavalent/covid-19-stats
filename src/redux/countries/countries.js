/* eslint-disable camelcase */
import axios from 'axios';

const ADD_COUNTRY = 'covid-19-stats/countries/ADD_COUNTRY';
const REMOVE_ALL_COUNTRIES = 'covid-19-stats/countries/REMOVE_ALL_COUNTRIES';

export const initialState = {
  today_confirmed: 0,
  today_new_deaths: 0,
  today_new_recovered: 0,
  today_recovered: 0,
  countries: {},
  europe: {
    today_confirmed: 0,
    today_new_deaths: 0,
    today_new_recovered: 0,
    today_recovered: 0,
    countries: {},
  },
  africa: {
    today_confirmed: 0,
    today_new_deaths: 0,
    today_new_recovered: 0,
    today_recovered: 0,
    countries: {},
  },
  // australia: { countries: {} },
  asia: {
    today_confirmed: 0,
    today_new_deaths: 0,
    today_new_recovered: 0,
    today_recovered: 0,
    countries: {},
  },
  north_america: {
    today_confirmed: 0,
    today_new_deaths: 0,
    today_new_recovered: 0,
    today_recovered: 0,
    countries: {},
  },
  south_america: {
    today_confirmed: 0,
    today_new_deaths: 0,
    today_new_recovered: 0,
    today_recovered: 0,
    countries: {},
  },
  oceania: {
    today_confirmed: 0,
    today_new_deaths: 0,
    today_new_recovered: 0,
    today_recovered: 0,
    countries: {},
  },
};

export const addCountry = (payload) => ({
  type: ADD_COUNTRY,
  payload,
});

export const removeAllCountries = () => ({
  type: REMOVE_ALL_COUNTRIES,
});

export const getCountries = () => (dispatch) => {
  function getYesterdayDateString() {
    const date = new Date(new Date().getTime() - 24 * 60 * 60 * 1000);
    const mm = date.getMonth() + 1; // getMonth() is zero-based
    const dd = date.getDate();

    return [date.getFullYear(), (mm > 9 ? '' : '0') + mm, (dd > 9 ? '' : '0') + dd].join('-');
  }

  (async () => {
    try {
      const yesterdayDateString = getYesterdayDateString();
      let covidCountriesResponse = await axios.get(`https://api.covid19tracking.narrativa.com/api/${yesterdayDateString}`); // await axios.get('https://api.covid19tracking.narrativa.com/api/countries');
      covidCountriesResponse = covidCountriesResponse.data.dates[yesterdayDateString];
      const countriesResponse = await axios.get('https://restcountries.com/v3.1/all');
      Object.entries(covidCountriesResponse.countries)
        .map(([id, country]) => ({ id, ...country }))
        .forEach(({
          id,
          name,
          today_confirmed,
          today_new_deaths,
          today_new_recovered,
          today_recovered,
          regions,
        }) => {
          const country = countriesResponse.data.filter((country) => (
            country.name.common === name
            || (country.altSpellings && country.altSpellings.includes(name))
          ))[0];
          const newRegions = {};
          regions.forEach((region) => {
            newRegions[region.name] = {
              name: region.name,
              today_confirmed: region.today_confirmed,
              today_new_deaths: region.today_new_deaths,
              today_new_recovered: region.today_new_recovered,
              today_recovered: region.today_recovered,
            };
          });
          if (country) {
            dispatch(addCountry({
              id,
              name,
              today_confirmed,
              today_new_deaths,
              today_new_recovered,
              today_recovered,
              regions: newRegions,
              flag: country.flag,
              flagSVG: country.flags.svg,
              continent: country.continents[0],
            }));
          }
        });
    // eslint-disable-next-line no-empty
    } catch (err) {}
  })();
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ADD_COUNTRY:
      // eslint-disable-next-line no-case-declarations
      const result = {
        ...state,
        countries: { ...state.countries, [action.payload.id]: action.payload },
        europe: { ...state.europe, countries: { ...state.europe.countries } },
        africa: { ...state.africa, countries: { ...state.africa.countries } },
        // australia: {...state.australia },
        asia: { ...state.asia, countries: { ...state.asia.countries } },
        north_america: { ...state.north_america, countries: { ...state.north_america.countries } },
        south_america: { ...state.south_america, countries: { ...state.south_america.countries } },
        oceania: { ...state.oceania, countries: { ...state.oceania.countries } },
      };
      // eslint-disable-next-line no-case-declarations
      const continentUrl = action.payload.continent.replaceAll(' ', '_').toLowerCase();
      result[continentUrl].countries[action.payload.id] = action.payload;
      result[continentUrl].today_confirmed += action.payload.today_confirmed;
      result[continentUrl].today_new_deaths += action.payload.today_new_deaths;
      result[continentUrl].today_new_recovered += action.payload.today_new_recovered;
      result[continentUrl].today_recovered += action.payload.today_recovered;
      result.today_confirmed += action.payload.today_confirmed;
      result.today_new_deaths += action.payload.today_new_deaths;
      result.today_new_recovered += action.payload.today_new_recovered;
      result.today_recovered += action.payload.today_recovered;
      return result;
    case REMOVE_ALL_COUNTRIES:
      return initialState;
    default:
      return state;
  }
};

export default reducer;
