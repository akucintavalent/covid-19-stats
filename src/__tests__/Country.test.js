/* eslint-disable no-extend-native */
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import '@testing-library/jest-dom';
import Country from '../components/country/Country';
import store from '../redux/configureStore';
import { addCountry } from '../redux/countries/countries';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: () => ({ country: 'ukraine' }), // jest.fn().mockReturnValue,
}));

if (typeof String.prototype.replaceAll === 'undefined') {
  String.prototype.replaceAll = function replaceAll(match, replace) {
    return this.replace(new RegExp(match, 'g'), () => replace);
  };
}

test('country snapshot', () => {
  store.dispatch(addCountry({
    id: 'ukraine',
    name: 'Ukraine',
    today_confirmed: 123,
    today_new_deaths: 456,
    today_new_recovered: 12304,
    today_recovered: 1238,
    regions: {},
    flag: '',
    flagSVG: '',
    continent: 'Europe',
  }));
  const tree = renderer
    .create(
      <Provider store={store}>
        <BrowserRouter>
          <Country />
        </BrowserRouter>
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
