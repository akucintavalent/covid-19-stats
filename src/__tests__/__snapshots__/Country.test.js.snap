// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`country snapshot 1`] = `
<div>
  <div
    className="stats"
    onClick={[Function]}
    onKeyDown={[Function]}
    type="button"
  >
       
    <i
      className="fas fa-chevron-left"
    />
       
  </div>
  <div
    className="country-header"
  >
    <img
      alt="flag of Ukraine"
      className="flag"
      src=""
    />
    <div>
      <p
        className="title"
      >
        Ukraine
      </p>
      <p>
        covid-19 stats for yesterday
      </p>
      <br />
      <p>
        confirmed:
        123
      </p>
      <p>
        new deaths:
        456
      </p>
      <p>
        new recovered:
        12304
      </p>
      <p>
        recovered:
        1238
      </p>
    </div>
  </div>
  <div
    className="stats"
  >
    STATS BY REGION
  </div>
  <div
    className="items"
  >
    <div
      className="no-regions"
    >
      NO REGIONS AVAILABLE
    </div>
  </div>
</div>
`;
